File: ReverseBits/recursive.c
#include <stdio.h>
#include <stdint.h>

uint32_t _reverseBits(uint32_t n, int bits) {
    int half = bits >> 1;
    if (bits == 1) return n;
    return (_reverseBits(n & ((1<<half)-1), half) << half) | _reverseBits(n >> half, half);
}

uint32_t reverseBits(uint32_t n) {
    return _reverseBits(n, 32);
}

int main() {
    printf("%d\n", reverseBits(43261596) == 964176192);
    printf("%d\n", reverseBits(964176192) == 43261596);
    printf("%d\n", reverseBits(0) == 0);
    printf("%d\n", reverseBits(1) == 2147483648);
    printf("%d\n", reverseBits(4294967295) == 4294967295);
    return 0;
}

File: ReverseBits/recursive.py
class Solution(object):
    def _reverse(self, n, bits):
        if bits == 1:
            return n
        halfbits = bits >> 1
        return (self._reverse(n & ((1 << halfbits) - 1), halfbits) << halfbits) | self._reverse(n >> halfbits, halfbits)

    def reverseBits(self, n):
        return self._reverse(n, 32)


def main():
    print(Solution().reverseBits(43261596) == 964176192)
    print(Solution().reverseBits(964176192) == 43261596)
    print(Solution().reverseBits(0) == 0)
    print(Solution().reverseBits(1) == 2147483648)
    print(Solution().reverseBits(4294967295) == 4294967295)

if __name__ == '__main__':
    main()
